### ISSUE TITLE
np.negative not deterministic, returning incorrect results

### ISSUE THREAD

**Issue by kevincmann** [numpy/numpy#11072]
I am using Xubuntu Linux 16.04 (python and numpy versions are included below). Running the following code:
```
import numpy as np
import sys

print('python:', sys.version)
print('numpy:', np.version.version)

b = np.array([1, 1, 1, 1, 1], dtype=float)
b_neg = (b < 0)

## uncomment the next line and it might behave
# print(b)
#

b = np.negative(b, where=(b_neg))

print(b)
```

It will print varying, similar, incorrect outputs, two of which from my computer are:
```
$ python main.py
python: 3.6.4 |Anaconda, Inc.| (default, Jan 16 2018, 18:10:19) 
[GCC 7.2.0]
numpy: 1.14.0
[0.00000000e+000 4.64016820e-310 5.08395840e-299 1.93322822e-260
 2.37151510e-322]

$ python main.py
python: 3.6.4 |Anaconda, Inc.| (default, Jan 16 2018, 18:10:19) 
[GCC 7.2.0]
numpy: 1.14.0
[0.00000000e+000 4.65820516e-310 5.08395840e-299 1.93322822e-260
 2.37151510e-322]
```

Uncommenting `print(b)` as noted in the code above fixes the problem in at least one version of python:
```
$ python main.py
python: 3.6.4 |Anaconda, Inc.| (default, Jan 16 2018, 18:10:19) 
[GCC 7.2.0]
numpy: 1.14.0
[1. 1. 1. 1. 1.]
[1. 1. 1. 1. 1.]
```

Other versions are still behaving strangely (even with the `print(b)` line uncommented):
```
$ python main.py
('python:', '2.7.13 |Anaconda custom (64-bit)| (default, Dec 20 2016, 23:09:15) 
[GCC 4.4.7 20120313 (Red Hat 4.4.7-1)]')
('numpy:', '1.10.4')
[ 1.  1.  1.  1.  1.]
[ 0.  0.  0.  0.  0.]
```

```
$ python3 main.py
python: 3.5.2 (default, Nov 23 2017, 16:37:01) 
[GCC 5.4.0 20160609]
numpy: 1.11.0
[ 1.  1.  1.  1.  1.]
[ 0.  0.  0.  0.  0.]
```

After updating to the latest version, it is still occurring:

With `print(b)` commented out:
```
$ python main.py
python: 3.6.5 |Anaconda custom (64-bit)| (default, Apr 29 2018, 16:14:56) 
[GCC 7.2.0]
numpy: 1.14.2
[ 4.68703220e-310 -8.44120994e-121  6.90939306e-310  6.90938300e-310
  5.34349229e+009]
```
With `print(b)` uncommented:
```
$ python main.py
python: 3.6.5 |Anaconda custom (64-bit)| (default, Apr 29 2018, 16:14:56) 
[GCC 7.2.0]
numpy: 1.14.2
[1. 1. 1. 1. 1.]
[1. 1. 1. 1. 1.]
```

**Comment by mhvk** [comment]
@kevincmann - it is not `np.negative` that is odd; this happens with all functions. And even though the output looks strange, it is actually expected: with the `where` argument, the output *only* gets set when the boolean is `True`. Normally, this is used with an `out` argument that already is filled, but here you do not pass in an `out` and thus an empty array is created. Since `where` is `False` everywhere, effectively you just get back the empty array. Now, the reason that the different versions and the print statement influence the result is just that those affect which memory is used for the empty array; apparently, with the print statement, you get a piece of memory that was previously initialized.

Does this make sense? Can you think of a way in which we could document this more clearly, so that you would not have been confused?

**Comment by eric-wieser** [comment]
Current docs are:

> ### where
> 
> > New in version 1.7.
> > 
> > Accepts a boolean array which is broadcast together with the operands. Values of True indicate to calculate the ufunc at that position, values of False indicate to leave the value in the output alone. This argument cannot be used for generalized ufuncs as those take non-scalar input.

I think it could be made clearer what `leave the value in the output alone` means here.

**Comment by mhvk** [comment]
Yes, I agree. @kevincmann - could you think what would have helped *you*? Ideally, it is just a few words, and I fear that if @eric-wieser or I write them they would clarify it for us but not for someone who has not been developing numpy! (If you're up to it, even make a PR!)

**Comment by mattip** [comment]
Perhaps

> If out is missing or None, a return value is created with uninitialized values.  The `ufunc` is applied, if `where` is used the `ufunc` will be applied where `where` is `True`. 

> If `out` is supplied and not None, the `ufunc` modifies it inplace, and the modified `out` array is returned. Note that the return value may have uninitialized values for `False` values in `where`

**Comment by kevincmann** [comment]
@mattip It is close.

I think if the second sentence in the second note were to be moved to the first note, then both would be accurate. 

In the first note, in the case when `where` is used, the second sentence of the second note applies, viz. there will be random values wherever `where` is not `True`. And in the case when `where` is not used, all (uninitialized) values are modified to correspond with the input with the expected changes (i.e. y = -x), so there will not be uninitialized values returned.

In the second note, when `out` is supplied and not `None`, there have not been uninitialized values returned in my testing. The `return` value of the function has been identical to the supplied `out` argument.

The code snippet from my previous comment shows the pertinent permutations that I have been using to form conclusions. `1` and `3` are relevant to your first note, and `2` and `4` are relevant to your second note.

**Comment by eric-wieser** [comment]
> And in the case when where is not used,

I don't think we need to call this out. `where` defaults to `True`, so this is already covered by the phrase "wherever where is not True"

**Comment by eric-wieser** [comment]
@kevin: Expressing mhvk's summary in code, the code path you care about is equivalent to:

```python
if out is None:
    out = np.empty(...)  # with correct type / shape etc
out[where] = operation(x, y)[where]
return out
```
